Remove FutureMessage from sender TensorPipeAgent

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* #50026 Remove FutureMessage from RPC request callback logic
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* **#50024 Remove FutureMessage from sender TensorPipeAgent**
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25753386](https://our.internmc.facebook.com/intern/diff/D25753386)

mrshenli committed:
Remove FutureMessage from sender TensorPipeAgent

[ghstack-poisoned]

mrshenli committed:
Update on "Remove FutureMessage from sender TensorPipeAgent"


Differential Revision: [D25753386](https://our.internmc.facebook.com/intern/diff/D25753386)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -273,8 +273,9 @@ class TensorPipeAgent : public RpcAgent {
-  struct AtomicFutureMessage {
-    FutureMessage futMsg;
+  struct AtomicJitFuture {
+    std::shared_ptr<JitFuture> jitFuture =
+        std::make_shared<JitFuture>(at::AnyClassType::get());
```
This change (replacing a value with a shared_ptr to a value) is undoing a slight optimization I had added: instead of having two nested shared_ptrs (hence two allocations, and two separate control blocks, with two reference counts) the code was using a single allocation (for the AtomicFutureMessage) with a single control block, and was relying on the shared_ptr "aliasing" constructor to extract a shared_ptr to the FutureMessage field that however still used the former control block.

The merits of this optimization are arguable, and it also has downsides (it keeps the atomic_flag alive even once it's not needed anymore), thus I'm fine to remove it, I just wanted to make sure this was intentional.

mrshenli replied:
This wasn't intentional. Let me add that to my todo list, and see if I can bring it back in a followup PR.

mrshenli committed:
Update on "Remove FutureMessage from sender TensorPipeAgent"


Differential Revision: [D25753386](https://our.internmc.facebook.com/intern/diff/D25753386)

[ghstack-poisoned]

mrshenli committed:
Update on "Remove FutureMessage from sender TensorPipeAgent"


Differential Revision: [D25753386](https://our.internmc.facebook.com/intern/diff/D25753386)

[ghstack-poisoned]
